openduck-py/openduck_py/response_agent.py

```diff
@@ -0,0 +1,381 @@
+        i = 0
+        while i < len(audio_16k_np):
+            upper = i + self.vad.window_size
+            if len(audio_16k_np) - self.vad.window_size < upper:
+                upper = len(audio_16k_np)
+            audio_16k_chunk = audio_16k_np[i:upper]
+            vad_result = self.vad(audio_16k_chunk)
+            if vad_result:
+                # TODO (Matthew): Can we send telemetry via an API instead of saving to a database?
+                async with SessionAsync() as db:
+                    if "end" in vad_result:
+                        print("end of speech detected.")
+                        self.time_of_last_activity = time()
+                        await log_event(db, self.session_id, "detected_end_of_speech")
+                        if self.response_task is None or self.response_task.done():
+                            self.response_task = asyncio.create_task(
+                                self.start_response(
+                                    np.concatenate(self.audio_data),
+                                )
+                            )
+                        else:
+                            print("already responding")
+                    if "start" in vad_result:
+                        print("start of speech detected.")
+                        self.time_of_last_activity = time()
+                        await log_event(db, self.session_id, "detected_start_of_speech")
+                        if self.response_task and not self.response_task.done():
+                            if self.is_responding:
+                                await log_event(
+                                    db, self.session_id, "interrupted_response"
+                                )
+                                self.interrupt(self.response_task)
+            i = upper
+
+    async def _generate_and_speak(
+        self,
+        db: SessionAsync,
+        t_whisper=None,
+        new_message=None,
+        system_prompt=None,
+        chat_model=CHAT_MODEL_GPT4,
+    ):
+        if system_prompt is None:
+            system_prompt = prompt(
+                f"most-interesting-bot/{self.system_prompt}", self.context
+            )
+        system_prompt = {
+            "role": "system",
+            "content": system_prompt,
+        }
+        # new_message = {"role": "user", "content": transcription}
+
+        chat = (
+            await db.execute(
+                select(DBChatHistory).where(DBChatHistory.session_id == self.session_id)
+            )
+        ).scalar_one_or_none()
+        if chat is None:
+            chat = DBChatHistory(
+                session_id=self.session_id,
+                history_json={"messages": [system_prompt]},
+            )
+            db.add(chat)
+        messages = chat.history_json["messages"]
+        if new_message is not None:
+            messages.append(new_message)
+
+        # NOTE(zach): retries
+        response = None
+        for _retry in range(3):
```

Comment:
not related to this PR but just wondering if the retries help? does it happen often that the openai API fails but then works again on retry? or is it a moderations thing?

Reply:
I'm not sure tbh, but I implemented this because things were flagged as moderation violations that clearly shouldn't have been